pkg/storage/mvcc.go

```diff
@@ -3595,7 +3596,12 @@ func MVCCPredicateDeleteRange(
 
 	var runStart, runEnd roachpb.Key
 
-	buf := make([]roachpb.Key, rangeTombstoneThreshold)
+	// buf holds keys that we might need to issue point deletes
+	// for. We copy the keys using keyAlloc, truncating keyAlloc
+	// if we don't send the point deletes and creating a new
+	// keyAlloc if we do send the point deletes.
+	var keyAlloc bufalloc.ByteAllocator
```

Comment:
Also, I understand we can truncate the byteAllocator if we issue a range tombstone -- those point keys will never get used again. 

Why don't you truncate the buffer after the point key flush as well? We truncate the buffer in both cases.

Reply:
Here "truncate" means "possibly reuse the underlying capacity.  But in the case of issuing the point deletes, that underlying capacity is being used for the Key's for the deletes we've issued and if we reuse the memory then we might end up issuing logical operations for the wrong key because of the memory overwrite.

Reply:
> it looks like the ByteAllocator has no unit tests and is currently only used to generate workloads. Before using this package in 23.2/1 for C2C, we ought to unit test this package. I am happy to help with this effort in a separate PR.

It is used in a few more places than that:

- `pkg/kv/kvserver/rangefeed/catchup_scan.go`
- `pkg/kv/kvserver/gc/gc.go`
- `pkg/sql/rowcontainer/disk_row_container.go`
- A few places in changefeeds

I agree some unit tests for this package would be prudent, but we definitely aren't the first production user of this.

Reply:
> But in the case of issuing the point deletes, that underlying capacity is being used for the Key's for the deletes we've issued and if we reuse the memory then we might end up issuing logical operations for the wrong key because of the memory overwrite.

aha. what I initially missed is why it's ok to _always_ truncate the buffer via `buf := buf[:0]`, regardless of a point or range key flush: the `buf` stores  _pointers_ to `roachpb.keys` since they keys are actually arrays of bytes. In `mvccPutInternal( buf[i],...)`, we're passing a _pointer_ to a key stored in `keyAlloc`.  It's fine to truncate the buf since `mvccPutInternal( buf[i],...)` has the right address to the key to delete. 

I think it may be worth explaining in the docstring why you only truncate after range deletes and why its safe to always truncate buf.

> [byteAllocator] is used in a few more places than that:
LOL, if I wait _several_ seconds, goland eventually does find all those references.